src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs

```diff
@@ -63,6 +63,13 @@ internal partial class AnalysisState
         private readonly ObjectPool<HashSet<CompilationEvent>> _compilationEventsPool;
         private readonly HashSet<CompilationEvent> _pooledEventsWithAnyActionsSet;
 
+        // Create static pools for heavily allocated per-analyzer state objects - this helps in reducing allocations across CompilationWithAnalyzer instances.
+        private const int PoolSize = 1000;
+        private static readonly ObjectPool<AnalyzerStateData> s_analyzerStateDataPool = new ObjectPool<AnalyzerStateData>(() => new AnalyzerStateData(), PoolSize);
```

Comment:
one of problem of using the pool I saw was that, most of time, it is fine, but let's say, just one time, there was a lot of elements to go in to object in the pool which makes the object size to blow up. if we never truncate it, then this big object just lives forever.

for example, for things like dictionary, we should have max size like 3000 items or something like that, and if dic's capacity is bigger than 3000, it should either let that dic go, or shrink it down.